### PR TITLE
[UserMenu] Create a new Item component that contains the user menu changes

### PR DESCRIPTION
**Pull request by zakwarsame**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/polaris-summer-editions/issues/785
Fixes https://github.com/Shopify/core-workflows/issues/1029
Fixes https://github.com/Shopify/polaris-summer-editions/issues/789

- The user menu's designs for uplift have been updated as follows:

Image

### WHAT is this pull request doing?

- This PR breaks away from the `ActionList`'s `Item` being used in the Menu item. Instead we create a custom `MenuItem` that can easily support features like `indentation` and removing constraints on the `avatar`.
- If a consumer wants to use this `MenuItem`, they provide a `topSection` prop that's passed to the UserMenu. The prop should specify to values: `indentItems` and `items`
- The UserMenu passes this prop down to `Menu` which applies these changes by rendering the new `MenuItem`
- The regular `actions` normally passed to `UserMenu` is not affected, thus anyone using `UserMenu` today does not need to make any changes.

A follow up to this PR is required after it's released, opened here: https://github.com/Shopify/web/pull/98734

### How to 🎩

 - Non-Plus store: https://admin.web.menu.zakaria-warsame.us.spin.dev/store/shop1
- Plus store: https://admin.web.menu.zakaria-warsame.us.spin.dev/store/shop2

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
~-[] Updated the component's `README.md` with documentation changesv~
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/TopBar/components/Menu/Menu.tsx [hunk 1]
```diff
@@ -32,8 +37,13 @@ export interface MenuProps {
   customWidth?: string;
   /** A boolean property indicating whether the menu is being used as a user menu */
   userMenu?: boolean;
+  /** Whether to indent the menu items, or not */
+  indent?: boolean;
 }
 
+const STORE_LIST_SECTION_ID = 'Stores';
+const OTHER_STORES_ID = 'otherStores';
```
Comment (zakwarsame) [inline polaris-react/src/components/TopBar/components/Menu/Menu.tsx]: The `Switch Organization` button has an `id` of otherStores. Assume variables with this name to be that item.

File: polaris-react/src/components/TruncateText/TruncateText.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import React, {useRef, useState} from 'react';
```
Comment (zakwarsame) [inline polaris-react/src/components/TruncateText/TruncateText.tsx:1]: Moving this component to the top-level since multiple components are now using it.

Comment (henryyi) [reply polaris-react/src/components/TruncateText/TruncateText.tsx:1]: I'll let the Polaris folk chime in but I'm not sure if this makes sense as a top level component named a such since it seems like it would be a replacement for `<Text truncate />`.
Maybe we can rename it to like: `TextWithTooltipWhenTruncated` but that's a bit wordy 😛

File: polaris-react/src/components/TopBar/components/Menu/Menu.tsx (hunk 1 above)
Comment (henryyi) [reply polaris-react/src/components/TopBar/components/Menu/Menu.tsx]: Having these ids hardcoded is a bit strange IMO. The Polaris component shouldn't care about the concept of stores.

File: polaris-react/src/components/TopBar/components/Menu/Menu.tsx [hunk 3]
```diff
@@ -93,15 +120,141 @@ export function Menu(props: MenuProps) {
...
+    ? storeListSection?.items.map((item, index) => {
+        const {itemMarkup, isOtherStore} = generateMenuItem(item, index);
+        if (isOtherStore) {
+          otherStoresMarkup = (
+            <Box width="100%" paddingInlineEnd="2" insetInlineEnd="2">
+              {itemMarkup}
+            </Box>
+          );
+          return null;
+        }
```
Comment (henryyi) [inline polaris-react/src/components/TopBar/components/Menu/Menu.tsx]: Can you clarify why "other store" need unique handling? I feel like this is too specific to the admin.

File: polaris-react/src/components/TopBar/components/UserMenu/UserMenu.tsx [hunk 4]
```diff
@@ -34,6 +34,8 @@ export interface UserMenuProps {
   customActivator?: React.ReactNode;
   /** A width value that customizes the width of the user menu */
   customWidth?: string;
+  /** Whether to indent the menu items, or not */
+  indent?: boolean;
```
Comment (henryyi) [inline polaris-react/src/components/TopBar/components/UserMenu/UserMenu.tsx]: It's a bit strange to set it here since it seems like it would apply to the whole menu and not just one section of it.
Perhaps we are better of specifying it as part of the actions.
My suggestion would be to modify the `actions`  prop to something unique to `UserMenu`. Maybe something like this:

1. Update ActionList item to allow for a `readOnly` prop so that we can render anytime as non-interactive

2. change to `actions` prop here to something that overrides `ActionListSection` and add a prop for indent:

```typescript
interface UserMenuAction extends ActionListSection {
    indent?: boolean;
}

actions: UserMenuAction[]
```

Comment (henryyi) [reply polaris-react/src/components/TopBar/components/UserMenu/UserMenu.tsx]: I guess we need at least one item that isn't indented at the bottom. In that case perhaps we can include the `indent` per item. Or we can have "unseparated" sections to make it easer:
```typescript
interface UserMenuAction extends ActionListSection {
    indent?: boolean;
    seamless?: boolean;
}
```

basically it would be used like:
```typescript
[
  {
    name: 'store-list',
    indent: true,
    items: [
      {
        name: 'current shop',
        readOnly: true,
        active, suffix, etc
      },
      ...otherItems,
    ]
  } ,
  {
    name: 'switch-org',
    seamless: true,
    items: [ switcheOrgButton]
  },
  ...otherSections
]
```

**Comment by zakwarsame** [comment]
@henryyi Please see this thread for why we're going with this approach: https://shopify.slack.com/archives/C4Y8N30KD/p1689780576017759
I think ultimately we'll want to pass in a prop to `ActionList`'s item and make our customizaitons there. However, the approach discussed for now was to conditionally our custom store section while leaving the User Menu unaffected.

File: polaris-react/src/components/TopBar/components/Menu/Menu.tsx (hunk 1 above)
Comment (zakwarsame) [reply polaris-react/src/components/TopBar/components/Menu/Menu.tsx]: I can understand eventually replacing `otherStores` (which is used to avoid the indentation for now) with item specific props, but how can we extract `Stores` without specifying it's ID? 
If we choose the first item without this ID, we risk affecting sections without the store list. If you can think of a better way to handle that, please let me know.

**Comment by zakwarsame** [comment]
@henryyi I've refactored the code so that we pass in a `topSection` prop with both the `indentItems` and regular `items` pre-separated from the consumer. That way Polaris is agnostic to the admin. Thanks a lot for the feedback!

File: polaris-react/src/components/TopBar/components/UserMenu/UserMenu.tsx (hunk 4 above)
Comment (zakwarsame) [reply polaris-react/src/components/TopBar/components/UserMenu/UserMenu.tsx]: I like the wrapper having to indent the section rather than individual items. Let me know what you think about the new approach

File: polaris-react/src/components/TopBar/components/Menu/Menu.tsx (hunk 3 above)
Comment (zakwarsame) [reply polaris-react/src/components/TopBar/components/Menu/Menu.tsx]: It was supposed to be the "Switch organization" button before the refactor.
